packages/playwright-core/src/server/frames.ts

```diff
@@ -738,18 +741,15 @@ export class Frame extends SdkObject {
         }
         if ((options as any).__testHookBeforeAdoptNode)
           await (options as any).__testHookBeforeAdoptNode();
+        const handle = result.asElement() as dom.ElementHandle<Element>;
         try {
-          const handle = result.asElement() as dom.ElementHandle<Element>;
-          const adopted = await handle._adoptTo(await this._mainContext());
-          return adopted;
+          return await handle._adoptTo(await frame._mainContext());
         } catch (e) {
-          // Navigated while trying to adopt the node.
-          if (js.isJavaScriptErrorInEvaluate(e))
-            throw e;
-          result.dispose();
+          return continuePolling;
+        } finally {
+          result?.dispose();
```

Comment:
Should not call `dispose` in the case of success, because you could dispose the returned handle (you should be able to trigger that with any selector engine that requires main world).